File: 4/largestpalprod.cpp
#include <iostream>
#include <cstdint>

using namespace std;

bool isPalindrome(uint64_t number) {
	string s = std::to_string(number);
	for (auto i = 0; i < s.length()/2; i++){
		if (s[i] != s[s.length()-1-i]) {
			return false;
		}
	}
	return true;
}

uint64_t largestPalindromeProduct() {
	uint64_t largestPalindromeProduct = 0;
	for (uint64_t a = 100; a <= 999; a++) {
		for (uint64_t b = 100; b <= 999; b++) {
			if (a > b) {
				continue;
			}
			uint64_t product = a * b;
			if (isPalindrome(product) && product > largestPalindromeProduct) {
				largestPalindromeProduct = product;
			}
		}
	}
	return largestPalindromeProduct;
}

int main() {
	uint64_t expectedAnswer = 906609;
	uint64_t answer = largestPalindromeProduct();
	if (answer == expectedAnswer) {
		cout << "PASSED" << endl;
	} else {
		cout << "FAILED (expected " << expectedAnswer << ", got " << answer << ")" << endl;
	}
}

File: 4/largestpalprod.go
package main

import (
	"fmt"
	"strconv"
)

func isPalindrome(number uint64) bool {
	var s string = strconv.FormatUint(number, 10)
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}

func largestPalindromeProduct() uint64 {
	var a uint64
	var b uint64
	var largestPalindromeProduct uint64 = 0
	for a = 100; a <= 999; a++ {
		for b = 100; b <= 999; b++ {
			var product uint64 = a * b
			if isPalindrome(product) && product > largestPalindromeProduct {
				largestPalindromeProduct = product
			}
		}
	}
	return largestPalindromeProduct
}

func main() {
	var expectedAnswer uint64 = 906609
	var answer uint64 = largestPalindromeProduct()
	if answer == expectedAnswer {
		fmt.Println("PASSED")
	} else {
		fmt.Println("FAILED (expected " + strconv.FormatUint(expectedAnswer, 10) + ", got " + strconv.FormatUint(answer, 10) + ")")
	}
}
